include/deal.II/fe/fe_enriched.h

```diff
@@ -644,6 +648,299 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * </tr>
+   * <tr>
+   * <td>1</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Q<2>(1)-FE_Nothing<2>()]</code></td>
+   * </tr>
+   * <tr>
+   * <td>2</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Q<2>(1)-FE_Q<2>(1)]</code></td>
+   * </tr>
+   * <tr>
+   * <td>3</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Nothing<2>()-FE_Q<2>(1)]</code></td>
+   * </tr>
+   * </table>
+   *
+   * The internal data members used by this class need to be available when the
+   * problem is solved. This can be ensured by declaring the object static,
+   * which then remains in the memory until the end of program. An alternative
+   * would be to use it as a data member of the containing class. Since vector
+   * of predicates and enrichment functions may not be available while
+   * constructing the Helper<dim>, a shared_ptr<Helper<dim>> can be used and
+   * constructed when the predicates and enrichment functions are
+   * available.
+   *
+   * Example usage:
+   * @code
+   * FE_Q<dim> fe_base(2);
+   * FE_Q<dim> fe_enriched(1);
+   * std::vector< predicate_function<dim> > predicates;
+   * std::vector< std::shared_ptr<Function<dim>> > enrichments;
+   * hp::FECollection<dim> fe_collection;
+   *
+   * Triangulation<dim>  triangulation;
+   * hp::DoFHandler<dim> dof_handler(triangulation);
+   *
+   * static ColorEnriched::Helper<dim> FE_helper(fe_base,
+   *                                             fe_enriched,
+   *                                             predicates,
+   *                                             enrichments);
+   * FE_helper.set(dof_handler);
+   * fe_collection = FE_helper.get_fe_collection();
+   * @endcode
+   */
+  template<int dim, int spacedim=dim>
+  struct Helper
+  {
+    /**
+     * Constructor for Helper class.
+     *
+     * @param fe_base A base FiniteElement
+     * @param fe_enriched An enriched FiniteElement
+     * @param predicates std::vector of predicates defining the sub-domains.
+     * <code>@p predicates[i]</code> returns true for a cell if it
+     * belongs to a sub-domain with index (i).
+     * @param enrichments std::vector of enrichment functions
+     */
+    Helper(const FE_Q<dim,spacedim> &fe_base,
+           const FE_Q<dim,spacedim> &fe_enriched,
+           const std::vector<predicate_function<dim,spacedim>> &predicates,
+           const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments);
+
+    /**
+     * Prepares @p a hp::DoFHandler object. The active FE indices of
+     * mesh cells are initialized to work with
+     * ColorEnriched::Helper<dim>::fe_collection.
+     *
+     * @param dof_handler a hp::DoFHandler<dim> object
+     */
+    void set(hp::DoFHandler<dim,spacedim> &dof_handler);
+
+    /**
+     * Returns a hp::FECollection<dim> object.
+     *
+     * @return hp::FECollection<dim>, a collection of finite elements
+     */
+    hp::FECollection<dim,spacedim> get_fe_collection() const;
```

Comment:
Is this intentionally that you return a copy of the FE here and not just a reference (or a `unique_ptr` of a clone)?

Reply:
This was intentional. I could have made it return a const reference but i didn't want to leak the implementation. I thought it would be more error safe as well this way.

Reply:
Copying a whole FE Collection object is a *really really* expensive operation. Let's try not to do that.

Reply:
What about a `std::shared_ptr`? That sounds like the best option if your class should still have (shared) ownership.

Reply:
Shared pointers only make sense if you expect that (i) it's unknown which side will keep the pointer for longer, or (ii) either side can modify the data.

I have not read through your code, so I don't know whether that's the case. But if all you want is to expose a member so others can see it -- like, for example, `DoFHandler::get_fe()` --, then returning a `const` reference is the way to go.